src/ip_notifier.cpp

```diff
@@ -99,6 +98,80 @@ void CFDispatchRetain(dispatch_queue_t q) { dispatch_retain(q); }
 void CFDispatchRelease(dispatch_queue_t q) { dispatch_release(q); }
 using CFDispatchRef = CFRef<dispatch_queue_t, CFDispatchRetain, CFDispatchRelease>;
 
+#if TARGET_OS_IPHONE || defined TORRENT_FORCE_IOS_IP_NOTIFIER
+CFRef<SCNetworkReachabilityRef> create_reachability(SCNetworkReachabilityCallBack callback
+	, void* context_info)
+{
+	TORRENT_ASSERT(callback != nullptr);
+
+	sockaddr_in addr = {};
+	addr.sin_len = sizeof(addr);
+	addr.sin_family = AF_INET;
+
+	CFRef<SCNetworkReachabilityRef> reach{SCNetworkReachabilityCreateWithAddress(nullptr
+		, reinterpret_cast<sockaddr const*>(&addr))};
+	if (!reach)
+		return CFRef<SCNetworkReachabilityRef>();
+
+	SCNetworkReachabilityContext context = {0, nullptr, nullptr, nullptr, nullptr};
+	context.info = context_info;
+
+	return SCNetworkReachabilitySetCallback(reach.get(), callback, &context)
+		? reach : CFRef<SCNetworkReachabilityRef>();
+}
+
+struct ip_change_notifier_impl final : ip_change_notifier
+{
+	explicit ip_change_notifier_impl(io_service& ios)
+		: m_ios(ios)
+	{
+		m_queue = dispatch_queue_create("libtorrent.IPChangeNotifierQueue", nullptr);
+		m_reach = create_reachability(
+			[](SCNetworkReachabilityRef /*target*/, SCNetworkReachabilityFlags /*flags*/, void *info)
+			{
+				auto obj = static_cast<ip_change_notifier_impl*>(info);
+				obj->m_ios.post([obj]() { if (obj->m_cb) obj->m_cb(error_code()); });
+			}, this);
+
+		if (!m_queue || !m_reach
+			|| !SCNetworkReachabilitySetDispatchQueue(m_reach.get(), m_queue.get()))
+			cancel();
+	}
+
+	// noncopyable
+	ip_change_notifier_impl(ip_change_notifier_impl const&) = delete;
+	ip_change_notifier_impl& operator=(ip_change_notifier_impl const&) = delete;
+
+	~ip_change_notifier_impl() override
+	{ cancel(); }
+
+	void async_wait(std::function<void(error_code const&)> cb) override
+	{
+		if (m_queue)
+			m_cb = std::move(cb);
+		else
+			m_ios.post([cb]()
+			{ cb(make_error_code(boost::system::errc::not_supported)); });
+	}
+
+	void cancel() override
+	{
+		if (m_reach)
+			SCNetworkReachabilitySetDispatchQueue(m_reach.get(), nullptr);
```

Comment:
an important property of boost.asio I/O objects is that one invocation of ``async_*()`` results exactly in one invocation of the callback function, under all circumstances. There may be some tricky edge cases around cancelation. Consider the case where we cancel the ip_notifier, but we're too late, a change has already happened and there's a notification queued up to be delivered to us. If we would unconditionally call post a call to ``m_cb`` here, when we cancel, it may be invoked twice.